internal/crypto/cms/signed_test.go

```diff
@@ -0,0 +1,424 @@
+			wantErr:  true,
+		},
+		{
+			name:     "with SHA1 hash",
+			filePath: "testdata/TimeStampTokenWithSignedAttributeSHA1.p7s",
+			wantErr:  true,
+		},
+		{
+			name:     "with invalid signing time",
+			filePath: "testdata/TimeStampTokenWithSigingTime.p7s",
+			wantErr:  true,
+		},
+		{
+			name:     "valid signing time",
+			filePath: "testdata/TimeStampTokenWithSigningTime.p7s",
+			wantErr:  false,
+		},
+		{
+			name:     "signing time before expected",
+			filePath: "testdata/TimeStampTokenWithSigningTimeBeforeExpected.p7s",
+			wantErr:  true,
+		},
+		{
+			name:     "timestamp token without signed attributes",
+			filePath: "testdata/TimeStampTokenWithoutSignedAttributes.p7s",
+			wantErr:  true,
+		},
+	}
+
+	for _, testcase := range testData {
+		t.Run(testcase.name, func(t *testing.T) {
+			// parse signed data
+			sigBytes, err := os.ReadFile(testcase.filePath)
+			if err != nil {
+				t.Fatal("failed to read test signature:", err)
+			}
+			signed, err := ParseSignedData(sigBytes)
+			if err != nil {
+				t.Fatal("ParseSignedData() error =", err)
+			}
+
+			// verify with no root CAs and should fail
+			// roots := x509.NewCertPool()
+			// certLen := len(signed.Certificates)
+			// if certLen > 0 {
+			// 	roots.AddCert(signed.Certificates[certLen-1])
+			// }
+			// opts := x509.VerifyOptions{
+			// 	Roots:       roots,
+			// 	KeyUsages:   []x509.ExtKeyUsage{x509.ExtKeyUsageTimeStamping},
+			// 	CurrentTime: time.Date(2024, 1, 9, 0, 0, 0, 0, time.UTC),
+			// }
+			err = signed.verifyAttributes(&signed.SignerInfos[0], [][]*x509.Certificate{signed.Certificates})
+			if testcase.wantErr && err == nil {
+				t.Errorf("ParseSignedData.Verify() error = %v, wantErr %v", err, true)
+			} else if !testcase.wantErr && err != nil {
+				t.Errorf("ParseSignedData.Verify() error = %v, wantErr %v", err, false)
+			}
+		})
+	}
+}
+
+func TestVerifyCorruptedSignedData(t *testing.T) {
+	// parse signed data
+	sigBytes, err := os.ReadFile("testdata/TimeStampToken.p7s")
+	if err != nil {
+		t.Fatal("failed to read test signature:", err)
+	}
+	signed, err := ParseSignedData(sigBytes)
+	if err != nil {
+		t.Fatal("ParseSignedData() error =", err)
+	}
+
+	// corrupt the content
+	signed.Content = []byte("corrupted data")
+
+	// verify with no root CAs and should fail
+	roots := x509.NewCertPool()
+	rootCABytes, err := os.ReadFile("testdata/GlobalSignRootCA.crt")
+	if err != nil {
+		t.Fatal("failed to read root CA certificate:", err)
+	}
+	if ok := roots.AppendCertsFromPEM(rootCABytes); !ok {
+		t.Fatal("failed to load root CA certificate")
+	}
+	opts := x509.VerifyOptions{
+		Roots:       roots,
+		KeyUsages:   []x509.ExtKeyUsage{x509.ExtKeyUsageTimeStamping},
+		CurrentTime: time.Date(2022, 1, 1, 0, 0, 0, 0, time.UTC),
+	}
+	if _, err := signed.Verify(opts); err == nil {
+		t.Errorf("ParseSignedData.Verify() error = %v, wantErr %v", err, true)
+	} else if _, ok := err.(VerificationError); !ok {
```

Comment:
I'm trying to understand, why would this test is expected to fail? Looks like we are adding a valid root cert into the `opts`?

Reply:
```go
signed.Content = []byte("corrupted data")
```
It has been tempered

Reply:
Ah, okay. So the comment at line 405 `// verify with no root CAs and should fail` is not accurate, we can remove it.

Reply:
Updated.